### DISCUSSION TITLE
[unreal]关于引擎类继承的构造函数和js初始化

### DISCUSSION THREAD

**Discussion by watsonsong** [Tencent/puerts#1436] [Q&A]
引擎类继承是不会调用到js自己的constructor的。因此所有的默认值都不会被赋初始值。一开始我是将这些纯js的对象都一起放到Constructor中进行初始化。
但是[文档](https://puerts.github.io/docs/puerts/unreal/uclass_extends#%E6%9E%84%E9%80%A0%E5%87%BD%E6%95%B0)建议的是：不能在该函数(Constructor)中做js的初始化工作，比如no-blueprint标注的变量的初始化。

如果是这个限制，引擎类继承中的纯js对象一般放到哪里初始化比较好。是这个引擎类自己的某个初始化回调(BeginPlay)这样的地方么？

**Comment by chexiongsheng** [comment]
BeginPlay可以

**Comment by watsonsong** [comment]
最近又测试了一下Constructor，发现一旦写了一个Constructor，不仅仅自己的代理蓝图上的数据会被清理掉，连所有派生代理蓝图的子类数据也会丢失。感觉要是一不小心谁写了一个Constructor可能引起一大堆的数据丢失。

衍生的另一个问题就是在属性同步的时候，因为蓝图好像没有办法设置REPNOTIFY_Always，一般是把数据（例如index: number）初始化成非法值(-1)，这样在客户端新加入的时候始终能收到一个OnRep_的事件。但发现首先是js自己的构造函数没法使用，其次Constructor中初始化成-1依然收不到数值。
目前唯一能工作的方法是，给代理蓝图派生一个子类蓝图，在子类蓝图中修改默认数据。但又会碰到上文说的一修改就重置。

**Comment by watsonsong** [comment]
想请教一下，如果一个引擎继承类不使用Constructor，有没有可能性让js的constructor能生效？

**Reply by chexiongsheng** [reply]
es5之前可以，没有class，都是用普通函数模拟，构造函数可以指定this来调用。
es6的class的构造函数不支持。
